smartsim/_core/_cli/build.py

```diff
@@ -331,84 +328,146 @@ def infer_torch_device(self) -> str:
-        """Check Python environment for TensorFlow installation"""
+    def check_py_tf_version(self, handle_conflict: bool = False) -> None:
+        """Check Python environment for a compatible TensorFlow installation"""
+        logger.info(f"Searching for a compatible TF install...")
+        self._check_py_package_version(
+            {"tensorflow": self.versions.TENSORFLOW},
+            install_on_absent=handle_conflict,
+            install_on_conflict=handle_conflict,
+        )
 
-        try:
-            if not self.build_env.check_installed(
-                "tensorflow", self.versions.TENSORFLOW
-            ):
-                msg = (
-                    f"TensorFlow {self.versions.TENSORFLOW} not installed in Python environment. "
-                    + f"Consider installing tensorflow=={self.versions.TENSORFLOW} with pip"
-                )
-                logger.warning(msg)
-            else:
-                logger.info(
-                    f"TensorFlow {self.versions.TENSORFLOW} installed in Python environment"
-                )
-        except SetupError as e:
-            logger.warning(str(e))
+    def _check_py_package_version(
+        self,
+        packages: t.Mapping[str, t.Optional[str]],
+        end_point: t.Optional[str] = None,
+        validator: t.Optional[t.Callable[[str, t.Optional[str]], bool]] = None,
+        install_on_absent: bool = False,
+        install_on_conflict: bool = False,
+    ) -> None:
+        missing: t.List[str] = []
+        conflicts: t.List[str] = []
+        to_uninstall: t.List[str] = []
+        to_install: t.List[str] = []
+        validator = validator or self.build_env.check_installed
+        verbose_info: t.Callable[[str], t.Any] = lambda s: ...
+        if self.verbose:
+            verbose_info = logger.info
+
+        for name, version in packages.items():
+            spec = f"{name}=={version}" if version else name
+            try:
+                if validator(name, version):
+                    # Installed at the correct version, nothing to do here
+                    verbose_info(f"{spec} already installed in environment")
+                else:
+                    # Not installed! Install preferred version or warn user
+                    if install_on_absent:
+                        verbose_info(
+                            f"Package not found: Queueing `{spec}` for install"
+                        )
+                        to_install.append(spec)
+                    else:
+                        missing.append(spec)
+            except VersionConflictError as e:
+                # Incompatible version found
+                if install_on_conflict:
+                    verbose_info(f"{e}: Queueing `{name}` for reinstall")
+                    to_uninstall.append(name)
+                    to_install.append(spec)
+                else:
+                    conflicts.append(spec)
+
+        if missing or conflicts:
+            indent = "\n    "
```

Comment:
If we separate out the output formatting, we could more easily write independent tests like `foo = _check_py_package_version(...); assert foo.conflicts; assert not foo.to_uninstall;`  (instead of capturing stdout and looking for output strings that will be less reliable)

Reply:
Brilliant idea!! Breaking logic out into a new helper fn and keeping this one around for formatting!